src/vm/ceeload.cpp

```diff
@@ -13234,6 +13239,283 @@ void ReflectionModule::ReleaseILData()
+    {
+        MethodTable* pMT;
+        TypeHandle* pDictSlots;
+    };
+    StackSArray<NewDictionary> dictionaryUpdates;
+
+#ifdef _DEBUG
+    auto expandPerInstInfos = [oldInfoSize, newInfoSize, &dictionaryUpdates](OBJECTREF obj, MethodTable* pMTKey, MethodTable* pMTToUpdate)
+#else
+    auto expandPerInstInfos = [newInfoSize, &dictionaryUpdates](OBJECTREF obj, MethodTable* pMTKey, MethodTable* pMTToUpdate)
+#endif
+    {
+        if (!pMTToUpdate->HasSameTypeDefAs(pMTKey))
+            return true;
+
+        _ASSERTE(pMTToUpdate != pMTToUpdate->GetCanonicalMethodTable() && pMTToUpdate->GetCanonicalMethodTable() == pMTKey);
+        _ASSERTE(pMTToUpdate->GetDictionarySlotsSize() == oldInfoSize);
+
+        TypeHandle* pInstOrPerInstInfo = AllocateNewTypeDictionaryForExpansion(pMTToUpdate, newInfoSize);
+
+        NewDictionary entry;
+        entry.pMT = pMTToUpdate;
+        entry.pDictSlots = pInstOrPerInstInfo;
+        dictionaryUpdates.Append(entry);
+
+        return true; // Keep walking
+    };
+
+    m_dynamicSlotsHashForTypes.VisitValuesOfKey(pCanonMT, expandPerInstInfos);
+
+    // Flush write buffers to ensure new dictionaries are fully writted and initalized before publishing them
+    FlushProcessWriteBuffers();
+
+    for (SArray<NewDictionary>::Iterator i = dictionaryUpdates.Begin(); i != dictionaryUpdates.End(); i++)
+    {
+        MethodTable* pMT = i->pMT;
+        TypeHandle** pPerInstInfo = (TypeHandle**)pMT->GetPerInstInfo()->GetValuePtr();
+        FastInterlockExchangePointer(pPerInstInfo + (pMT->GetNumDicts() - 1), i->pDictSlots);
+        _ASSERTE(pMT->GetDictionarySlotsSize() == newInfoSize);
+        _ASSERTE((TypeHandle*)pMT->GetPerInstInfo()[pMT->GetNumDicts() - 1].GetValue() == i->pDictSlots);
+    }
+
+    auto updateDependentDicts = [](OBJECTREF obj, MethodTable* pMTKey, MethodTable* pMTToUpdate)
+    {
+        if (pMTToUpdate->HasSameTypeDefAs(pMTKey))
+            return true;
+
+        MethodTable* pCurrentMT = pMTToUpdate->GetParentMethodTable();
+        while (pCurrentMT)
+        {
+            if (pCurrentMT->HasSameTypeDefAs(pMTKey))
+            {
+                DWORD dictToUpdate = pCurrentMT->GetNumDicts() - 1;
+                Dictionary* pUpdatedParentDict = pCurrentMT->GetPerInstInfo()[dictToUpdate].GetValue();
+                TypeHandle** pPerInstInfo = (TypeHandle**)pMTToUpdate->GetPerInstInfo()->GetValuePtr();
+                FastInterlockExchangePointer(pPerInstInfo + dictToUpdate, (TypeHandle*)pUpdatedParentDict);
+                _ASSERTE(pMTToUpdate->GetPerInstInfo()[dictToUpdate].GetValue() == pUpdatedParentDict);
+
+                return true; // Keep walking
+            }
+            pCurrentMT = pCurrentMT->GetParentMethodTable();
+        }
+
+        UNREACHABLE();
+    };
+
+    m_dynamicSlotsHashForTypes.VisitValuesOfKey(pCanonMT, updateDependentDicts);
+
+    // Ensure no other thread uses old dictionary pointers
+    FlushProcessWriteBuffers();
```

Comment:
I believe we shouldn't need this anymore, given that dictionary pointers are updated using FastInterlockExchangePointer, right? @davidwrighton

Reply:
I'm certain that interlocked updates protect dependent reads that see the update, but I'm not confident they ensure that all reading threads will have seen the update. FlushProcessWriteBuffers protects all reads. Its a subtly different semantic. I'll need to do some digging through the ARM and Intel documentation to do this analysis. That'll take some time.

Reply:
It's fine, we can just leave the flush call here.